### ISSUE TITLE
Internal Server Issue with spatie/laravel-permission

### ISSUE THREAD

**Issue by shsbd** [laravel/telescope#159]
Hello and thanks for the package. I have a project integrated with [https://github.com/spatie/laravel-permission](https://github.com/spatie/laravel-permission)

After installing package to Telescope, I can see there is Internal Server Error causing. Also, the permissions of the users are not showing. Contents which are wrapped with permission that shows blank screen instead of the contents. I am using Laravel's default **@can('')** for permission.

After uninstalling the Telescope package, application working again as expected. Need to know if there is any dependency or override issues happening which is causing the internal server error.

Previously I had used [https://github.com/barryvdh/laravel-debugbar](https://github.com/barryvdh/laravel-debugbar) which was running smoothly without causing any errors.

> Laravel Framework 5.7.11
> PHP 7.2.10
> MySQL  Ver 14.14 Distrib 5.7.23, for Win64 (x86_64)

|-------------------------------------
| Common Checks
|-------------------------------------
Running check 1/11: App key is set...  ✔
Running check 2/11: The correct PHP version is installed...  ✔
Running check 3/11: The database can be accessed...  ✔
Running check 4/11: All directories have the correct permissions...  ✔
Running check 5/11: The environment file exists...  ✔
Running check 6/11: The example environment variables are set...  ✔
Running check 7/11: Required locales are installed...  ✘
Running check 8/11: Maintenance mode is not enabled...  ✔
Running check 9/11: The migrations are up to date...  ✔
Running check 10/11: The required PHP extensions are installed...  ✔
Running check 11/11: The storage directory is linked...  ✔

|-------------------------------------
| Environment Specific Checks (local)
|-------------------------------------
Running check 1/4: Composer dependencies (including dev) are up to date...  ✔
Running check 2/4: Configuration is not cached...  ✔
Running check 3/4: Routes are not cached...  ✔
Running check 4/4: The example environment variables are up-to-date...  ✔

Thank you.

**Comment by turbo124** [comment]
I believe this is similar to other issues which have been raised and closed already.

#67
#109

I haven't been able to find the exact problem yet, but I believe the issue is that these classes have not been loaded prior to Telescope. These issues were not there prior to loading in telescope and go away when the telescope package is removed, so the issue is definitely with Telescope and the way it is positioning itself in the stack.

**Comment by themsaid** [comment]
What is the full stack trace of the exception?

**Comment by jamesggordon** [comment]
I have hit the same issue as @turbo124 mentioned in #67.

I have a Laravel 5.7.11 app with a custom user provider.

No Telescope everything works fine.

I add Telescope to composer.json and run composer update. I haven't run telescope:install yet, so there is no entry for the TelescopeServiceProvider in app.php.

Everything *still* works fine, **except** for config:cache, which fails with "Authentication user provider [foo] is not defined." What is really strange is that everything still works - I just can't cache the config file.

I'm commenting here because #67 was closed and feel like the underlying cause might be the same.

I will create a fresh Laravel app with a custom auth provider and see if the issue is reproducible that way.

**Comment by jamesggordon** [comment]
I have successfully reproduced #67 by creating a fresh Laravel app, writing a very basic custom user provider and adding Telescope to composer.json. Doing so results in an app that runs perfectly except that config:cache does not work.

I have added the repo here:

https://github.com/jamesggordon/telescope-issue-159

To reproduce the issue simply:

1. Clone the repo
2. composer install
3. php artisan config:cache (fails - "Authentication user provider [custom] is not defined.")
4. Remove telescope from composer.json
5. composer update
6. php artisan config:cache (works)

This is very, very strange indeed. You can comment out the TelescopeServiceProvider entry from app.php and remove the class from the Providers folder and the problem *still* occurs.

I know I haven't reproduced the issue that @shsbd originally posted about, but as mentioned earlier, it definitely feels to me like both issues are being caused by the same underying problem.

**Comment by themsaid** [comment]
Guys :) What is the stack trace of the exception you get?

**Comment by turbo124** [comment]
@themsaid please revisit #67 i included a stacktrace there..

if you see my post above, these providers are not being registered when telescope is enabled.

**Comment by themsaid** [comment]
Can someone who can replicate open a PR with a change that fixed it in their app?

**Comment by jamesggordon** [comment]
@themsaid Here is the stack trace of the exception that gets thrown at step 3 (php artisan config:cache) when Telescope has been added to composer.json:

    [2018-10-26 22:35:27] local.ERROR: Authentication user provider [custom] is not defined. {"exception":"[object] (InvalidArgumentException(code: 0): Authentication user provider [custom] is not defined. at /Users/james/dev/telescope-issue-159/vendor/laravel/framework/src/Illuminate/Auth/CreatesUserProviders.php:42)
    [stacktrace]
    #0 /Users/james/dev/telescope-issue-159/vendor/laravel/framework/src/Illuminate/Auth/AuthManager.php(121): Illuminate\\Auth\\AuthManager->createUserProvider('users')
    #1 /Users/james/dev/telescope-issue-159/vendor/laravel/framework/src/Illuminate/Auth/AuthManager.php(94): Illuminate\\Auth\\AuthManager->createSessionDriver('web', Array)
    #2 /Users/james/dev/telescope-issue-159/vendor/laravel/framework/src/Illuminate/Auth/AuthManager.php(68): Illuminate\\Auth\\AuthManager->resolve('web')
    #3 /Users/james/dev/telescope-issue-159/vendor/laravel/framework/src/Illuminate/Auth/AuthManager.php(292): Illuminate\\Auth\\AuthManager->guard()
    #4 /Users/james/dev/telescope-issue-159/vendor/laravel/framework/src/Illuminate/Support/Facades/Facade.php(223): Illuminate\\Auth\\AuthManager->__call('hasUser', Array)
    #5 /Users/james/dev/telescope-issue-159/vendor/laravel/telescope/src/Telescope.php(199): Illuminate\\Support\\Facades\\Facade::__callStatic('hasUser', Array)
    #6 /Users/james/dev/telescope-issue-159/vendor/laravel/telescope/src/Telescope.php(275): Laravel\\Telescope\\Telescope::record('exception', Object(Laravel\\Telescope\\IncomingExceptionEntry))
    #7 /Users/james/dev/telescope-issue-159/vendor/laravel/telescope/src/Watchers/ExceptionWatcher.php(46): Laravel\\Telescope\\Telescope::recordException(Object(Laravel\\Telescope\\IncomingExceptionEntry))
    #8 /Users/james/dev/telescope-issue-159/vendor/laravel/framework/src/Illuminate/Events/Dispatcher.php(360): Laravel\\Telescope\\Watchers\\ExceptionWatcher->recordException(Object(Illuminate\\Log\\Events\\MessageLogged))
    #9 /Users/james/dev/telescope-issue-159/vendor/laravel/framework/src/Illuminate/Events/Dispatcher.php(209): Illuminate\\Events\\Dispatcher->Illuminate\\Events\\{closure}('Illuminate\\\\Log\\\\...', Array)
    #10 /Users/james/dev/telescope-issue-159/vendor/laravel/framework/src/Illuminate/Log/Logger.php(210): Illuminate\\Events\\Dispatcher->dispatch('Illuminate\\\\Log\\\\...')
    #11 /Users/james/dev/telescope-issue-159/vendor/laravel/framework/src/Illuminate/Log/Logger.php(174): Illuminate\\Log\\Logger->fireLogEvent('error', 'Authentication ...', Array)
    #12 /Users/james/dev/telescope-issue-159/vendor/laravel/framework/src/Illuminate/Log/Logger.php(87): Illuminate\\Log\\Logger->writeLog('error', 'Authentication ...', Array)
    #13 /Users/james/dev/telescope-issue-159/vendor/laravel/framework/src/Illuminate/Log/LogManager.php(488): Illuminate\\Log\\Logger->error('Authentication ...', Array)
    #14 /Users/james/dev/telescope-issue-159/vendor/laravel/framework/src/Illuminate/Foundation/Exceptions/Handler.php(114): Illuminate\\Log\\LogManager->error('Authentication ...', Array)
    #15 /Users/james/dev/telescope-issue-159/app/Exceptions/Handler.php(37): Illuminate\\Foundation\\Exceptions\\Handler->report(Object(InvalidArgumentException))
    #16 /Users/james/dev/telescope-issue-159/vendor/nunomaduro/collision/src/Adapters/Laravel/ExceptionHandler.php(60): App\\Exceptions\\Handler->report(Object(InvalidArgumentException))
    #17 /Users/james/dev/telescope-issue-159/vendor/laravel/framework/src/Illuminate/Foundation/Console/Kernel.php(353): NunoMaduro\\Collision\\Adapters\\Laravel\\ExceptionHandler->report(Object(InvalidArgumentException))
    #18 /Users/james/dev/telescope-issue-159/vendor/laravel/framework/src/Illuminate/Foundation/Console/Kernel.php(124): Illuminate\\Foundation\\Console\\Kernel->reportException(Object(InvalidArgumentException))
    #19 /Users/james/dev/telescope-issue-159/artisan(37): Illuminate\\Foundation\\Console\\Kernel->handle(Object(Symfony\\Component\\Console\\Input\\ArgvInput), Object(Symfony\\Component\\Console\\Output\\ConsoleOutput))
    #20 {main}
    "}

**Comment by jamesggordon** [comment]
After additional investigation I have learned the following:

1. When you add Telescope to composer.json and `run composer install`, Telescope is installed and operational. That is, even though the assets haven't been published yet and `App\Providers\TelescopeServiceProvider` hasn't been created yet, Telescope **is** actually bound into the kernel and is looking for events to record.

2. When a command such as `php artisan config:clear` is run, it would appear that there are no events that Telescope wants to record and, as such, there are no issues.

3. However, when `php artisan config:cache` is run, it would appear that Telescope *does* want to record an event. However, when `Telescope::record()` is called, at around line 199 that method calls the `Auth::hasUser()` method. This results in the `AuthManager->createSessionDriver()` method getting called, which calls the `createUserProvider()` method from the `CreatesUserProviders` trait. However, at the point where `CreatesUserProviders->createUserProvider()` method is called, the `customProviderCreators[]` array, which is supposed to have an entry for the 'custom' user provider, is empty. Thus, it throws the exception shown above.

4. I added `echo count($this->customProviderCreators);` at the start of the `CreatesUserProviders->createUserProvider()` method and ran `php artisan config:clear`. As expected, there was one entry in the array (i.e. the 'custom' user provider). However, when I ran `php artisan config:cache`, there were zero entries in the array, which proves that the custom user provider hadn't been set up yet, which is what causes the exception.

So it would appear that when Telescope is installed, it is wired into the kernel and starts looking for events to record. However, it seems to be wired in to run **before** the custom user providers are set up, which is why it fails when it tries to record an event.

I don't know enough about the IOC container or the application bootstrapping to figure out how to load Telescope after the custom user provider has been created. However, until this issue is fixed, I don't see how Telescope can be working at all for *any* apps that are using custom user providers.

And just to re-iterate, I have created and replicated this issue with a fresh Laravel app, creating just one class, the `CustomUserProvider` class, configuring it in `auth.php` and bootstrapping it using the `boot()` method of the standard `App\Providers\AuthServiceProvider` class, exactly as per the documentation. Based on all that, I have reached the firm conclusion that this is a framework issue, not an application issue.

**Comment by turbo124** [comment]
So here is the workaround I have found which should resolve this issue for anyone using any custom implementations in any service provider.

In the Laravel documentation, it is suggested to place your custom closures inside the boot() method, you need to move these implementations into the register() method which telescope will pickup, so in my case where using a custom UserProvider:

I went from:

```
    public function boot()
    {
        $this->registerPolicies();

        Auth::provider('users', function ($app, array $config) {
            return new MultiDatabaseUserProvider($app['hash'], $config['model']);
        });
    }
```

to this:

```
    public function register()
    {
        Auth::provider('users', function ($app, array $config) {
            return new MultiDatabaseUserProvider($app['hash'], $config['model']);
        });
    }
```


My tests pass using this method, I am unsure if I am missing something here, but I think this is the only solution without having to make changes to the telescope package itself.

**Comment by jamesggordon** [comment]
Interesting. I tried your workaround and did worked for me. Nice one!

However, I would feel uneasy about this becoming the recommended solution for everybody, because the documentation on Service Providers does state that "within the register method, you should only bind things into the service container. You should never attempt to register any event listeners, routes, or any other piece of functionality within the register method. Otherwise, you may accidentally use a service that is provided by a service provider which has not loaded yet."

Hopefully @themsaid will get a chance to have another look at this one, because I reckon with his deep knowledge of the IOC container and the bootstrapping process he would be able to solve this one in a few minutes.

**Comment by derekmd** [comment]
"laravel/laravel" suggests ordering the app-level `AppServiceProvider` _then_ `AuthServiceProvider`:

https://github.com/laravel/laravel/blob/b98d49ebb31e3875aef2a5696c5cbc7bf58883f2/config/app.php#L170-L171

If `AuthServiceProvider` is registering a custom user provider, I suggest **reordering to the opposite**.

```
'providers' => [
// ...
    App\Providers\AuthServiceProvider::class,
    App\Providers\AppServiceProvider::class,
```

This gives Telescope access to gate policies when registering.

e.g., I chose to disable Telescope in production (for now) by turning off package auto-discovery.

```php
namespace App\Providers;

class AppServiceProvider extends ServiceProvider
{
    public function boot()
    {
        Horizon::auth(function ($request) {
            return optional($request->user())->isAdmin();
        });
    }

    public function register()
    {
        if ($this->app->environment('testing', 'local', 'staging')) {
            $this->app->register(\BeyondCode\DumpServer\DumpServerServiceProvider::class);
            $this->app->register(\Laravel\Dusk\DuskServiceProvider::class);
        }

        if ($this->app->environment('local', 'staging')) {
            $this->app->register(\Laravel\Telescope\TelescopeServiceProvider::class);
            $this->app->register(\App\Providers\TelescopeServiceProvider::class);
        }
    }
}
```

In bootstrap/cache/services.php (generated by `php artisan cache:clear`), `App\Providers\AuthServiceProvider` had to be ordered above this `AppServiceProvider`.

**Comment by jamesggordon** [comment]
Thank you @derekmd for the suggestions.

I tried re-ordering the providers in `app.php` (which dictates their order in `services.php`) but unfortunately it didn't solve the problem for me. I also tried registering the `TelescopeServiceProvider` classes in the `register()` method of the `AppServiceProvider` class as above. Unfortunately that did not solve the problem either. The only thing that has solved the problem for me thus far is registering the custom user provider in the `register()` method of the `AuthServiceProvider`, but as this goes in contradiction with the docs, I have not implemented that solution.

For reference, I am using a fresh Laravel project with a minimal custom user provider class just for testing this issue. https://github.com/jamesggordon/telescope-issue-159

Also, just as an aside, `php artisan cache:clear` clears the application cache, not the compiled cache. You want `php artisan clear-compiled` for that, which deletes the files in `bootstrap/cache` and they will be re-generated the next time the framework boots up.

**Comment by themsaid** [comment]
I still can't seem to replicate the issue and I feel like people here are discussing different things, one related to a package, another related to config:cache command error.

I'm closing this issue, please open a new issue with a steps to replicate in a fresh laravel installation.

I tried https://github.com/jamesggordon/telescope-issue-159 and everything seems to be working fine, so instead of sharing a repo please share some steps I can apply on a fresh installation to replicate the error.

Thanks everyone.

**Comment by jamesggordon** [comment]
Thanks for following @themsaid - I really appreciate it. I can also confirm that this issue has been resolved as at version 0.1.8 of Telescope.

I ran `php artisan config:cache` this morning before running `composer update` and could still replicate the issue. However, I then ran `composer update` and the Telescope version went from 0.1.3 to 0.1.8. Then, when I ran `php artisan config:cache`, the error that I was previously getting did not occur. So there definitely *was* an issue, but it has gone away, so all good! :-)
